### PR TITLE
GlobalTransform Rework

### LINKED ISSUE #2026: Reduce code duplication between Transform and GlobalTransform
## What problem does this solve or what need does it fill?

Currently, `Transform` and `GlobalTransform` have effectively the same interface, just as different components. They are implemented as completely different types. Code is duplicated between the two and there is no simple way to write functions that are generic over both.

## What solution would you like?

Make the definition of `GlobalTransform`:
```rust
#[repr(transparent)]
struct GlobalTransform(pub Transform);
```

## What alternative(s) have you considered?

Creating a trait implemented by both transform types. I don't think there's any real reason for this because the actual implementation details are *identical*, as far as I know. A trait would make sense if the implementation differed, but since they don't I believe composition makes more sense.

Another thing we could add is a `Deref<Target = Transform>` impl for `GlobalTransform`. This would get rid of some `global_transform.0` calls, but I think it's not worth it. Also, even though this is a common pattern in the rust ecosystem, it goes directly against the standard library's documentation for the intended usage of the `Deref` trait:

> Implementing Deref for smart pointers makes accessing the data behind them convenient, which is why they implement Deref. On the other hand, the rules regarding Deref and DerefMut were designed specifically to accommodate smart pointers. Because of this, Deref should only be implemented for smart pointers to avoid confusion.

## Additional context

This would be a breaking change, and would require a small modification to any code that uses global transforms.

The `#[repr(transparent)]` isn't strictly necessary, but there's no real reason not to add it.

I'm happy to write a PR for this if people agree that it's a good idea.

### PR DESCRIPTION
**Pull request by ManevilleF**
> Fixes #2026
> Conflicts with #4379

# Objective

Currently, `GlobalTransform` and `Transform` are absolutely identical, with duplicate fields, duplicate methods and constructors.
Removing the redundancy and making `Transform` the internal value of `GlobalTransform` would remove the duplicate code and restrict `GlobalTransform` to be a hierarchy component.

## Solution

`GlobalTransform` is now a tuple struct with an inner `Transform` value. It implements `Deref` towars its inner type allowing the use for the entire `Transform` API.
The inner type is not public, meaning that only `bevy_transform` can edit it. Enforcing the tacit **no touch** rule on that component.

## Notes

- Additionnaly I would like to remove `GlobalTransform` from every *bundle* and add it to entities through a system going through `Added<Transform>` artifacts.
- The `global_vs_local_transform` example can no longer work so I removed it (See #5139 )
- This MR can be ignored entirely if #4379 is merged

### CODE REVIEW

File: crates/bevy_transform/src/components/global_transform.rs [hunk 1]
```diff
@@ -24,249 +24,43 @@ use std::ops::Mul;
...
 }
 
-impl Default for GlobalTransform {
-    fn default() -> Self {
-        Self::identity()
+impl DerefMut for GlobalTransform {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
     }
 }
```
Comment (SkiFire13) [inline crates/bevy_transform/src/components/global_transform.rs:50]: Trait implementations are public, so even though the inner field is only `pub(crate)`, this implementation allows it to be modified by any crate.

**Review by james7132 — COMMENTED** [review]
I haven't fully read through the exact changes, but there's still the `std::mem::swap` loophole of mutating `GlobalTransform`. Otherwise this seems to be a shut case on preventing users from arbitrarily changing the transforms of everything.

**Comment by ManevilleF** [comment]
@james7132 The objective is mainly to remove redundancy and to avoid *accidental* misuse of `GlobalTransform` by editing its values, so I'm not worried by `mem::swap` or potential `EntityCommands::insert` to override the component.

But maybe we do want to allow the users to edit the `GlobalTransform` and then, implementing `DerefMut` does the trick

**Comment by HackerFoo** [comment]
This change requires that `Transform` and `GlobalTransform` have the same representation, which is problematic because they serve different purposes. `Transform` needs to map well to desired motions for animation and physics, and so rotation and scaling should be separate, as they are using the current SRT representation. `GlobalTransform` on the other hand is an accumulator for the chain of multiplied transforms. The problem (that #4379 solves) is that the SRT representation that `Transform` uses can't actually do that in the presence of intermediate nonuniform scale, and so gives the wrong result.

**Comment by alice-i-cecile** [comment]
My preference is for #4379 over this PR at this point :)

**Comment by ManevilleF** [comment]
This MR only removes code duplication and confusion on the two transform types.
I'm not making a stand on the inner type, quite the opposite. This lays groundwork for Global transform rework in the most uncontroversial way

**Comment by HackerFoo** [comment]
The PR converts `GlobalTransform` to the inner type in many places. This removes the opportunity to define methods and trait implementations specific to `GlobalTransform` (or restrict operations available on the inner type), and also makes it easier to confuse `GlobalTransform` with its' inner type, losing the information that it transforms from some local space to world space.

Doing this is the opposite of what I'd like to see, which is using types to avoid confusing different transforms, for example `Transform<B, C> * Transform<A, B> = Transfom<A, C>`. Transforms should be typed like any other function. This would have saved me a lot of time and frustration.

**Comment by ManevilleF** [comment]
> The PR converts `GlobalTransform` to the inner type in many places. This removes the opportunity to define methods and trait implementations specific to `GlobalTransform` (or restrict operations available on the inner type), and also makes it easier to confuse `GlobalTransform` with its' inner type, losing the information that it transforms from some local space to world space.

I convert to the inner type so that extraction stages no longer store a `GlobalTransform` wrapper but an actual `Transform`,  meaning *less* confusion between components and operational types imo. Ideally extraction types would store directly a `Mat4` or `Affine3A` but I'm just removing boilerplate here.

This MR only does this:
- A `Transform` is both as a component stores local transformation, and the transformation operational type.
- A `GlobalTransform` is only a component identifying world space transformation

> Doing this is the opposite of what I'd like to see, which is using types to avoid confusing different transforms, for example `Transform<B, C> * Transform<A, B> = Transform<A, C>`. Transforms should be typed like any other function. This would have saved me a lot of time and frustration.

I don't think making `Transform` generic is a good idea in term of ergonomics.

I'm closing this for now. I don't want this PR to shift focus from #4379

**Pull request closed without merging**